Update auth status when clicking validate for existing ems

jrafanie opened the pull request:
EDIT: 
**Issue**:  ems based workers no longer try to start if the ems's authentication was detected as previously bad, see #1691.  Finally, if the ems was unreachable and marked as bad, editing the ems in the UI and clicking "validate" doesn't "fix" the authentication's status, making it hard to make the ems's workers start.  The only **workaround** previously was to save invalid credentials, edit it again, and save the correct credentials.

https://bugzilla.redhat.com/show_bug.cgi?id=1044608

**Problem**:
The "workaround" above is horrible for users.

**Solution** validate button in UI should "revalidate" an unchanged ems (ie, when save is not enabled) and update it's authentication status so workers can start back up.

`authentication_check` defaults to verifying the credentials and saving the resulting status.

If you wanted to bypass the saving of the authentication's status, you can do this:
`authentication_check(:save => false)`.

The EMS edit UI code has been changed to no longer do error handling and instead call this method with `:save => false` or `:save => true` if the EMS's edit form has been changed or not.

This example code demonstrates how a client of this code would use these changes:
```ruby
    result, details = verify_ems.authentication_check(params[:type], :save => revalidating)
    if result
      add_flash(_("Credential validation was successful"))
    else
      add_flash(_("Credential validation was not successful: #{details}"), :error)
    end
```

jrafanie commented:
cc @h-kataria @Fryguy @blomquisg @mkanoor 

I really have no idea how to fix this without gutting just the "edit/update existing host/ems" forms.

jrafanie commented:
Additionally, I'm not sure how to handle the ssh mistmatch key logic that's being done in the UI.

Fryguy commented:
Note the UI validates the user/pass on a new record *before* it's saved to the database.

jrafanie commented:
Note, this is for existing records, not new ones

jrafanie commented on vmdb/app/controllers/ems_common.rb:
```diff
@@ -336,17 +340,16 @@ def update_button_validate
-    rescue StandardError => bang
-      add_flash("#{bang}", :error)
+
+    # validate button should say "revalidate" if the form is unchanged
```
@h-kataria What do you think of these UI changes?

The problem is if the current status of the creds is invalid and you edit the Ems, the validate button WITH NO CHANGES in the form, should revalidate the creds and UPDATE the status.  As soon as you make changes to the form, the validate button SHOULD NOT update the status and authentication (since that bypasses save).

jrafanie replied:
Note, I still don't know what to do with Host's update form validate button since that catches a ssh key mismatch exception.  Maybe I'll do that after getting this issue fixed since this is really a problem with the Ems having a bad auth status and the validate button doesn't do anything.

jrafanie commented:
@blomquisg I put together some changes that are now working when I manually tweak the my ems's status (by editing /etc/hosts so the ems is found/not found at will).  @Fryguy helped me clean some of it up.  I still have a few todos to fix this issue including writing tests for the new behavior of the mixin.

Did you have a BZ I can link to this issue?

Thanks.

Fryguy commented:
cc @AparnaKarve

jrafanie commented:
@h-kataria @AparnaKarve please review the changes to ems_common and authentication_mixin.
I moved the exception handling in the controller to the mixin and provided a way to tell `authentication_check` whether it should save the authentication or not: 

* Save when the form is unchanged and you click "validate" to "revalidate".  
* Don't save when the form is changed and you click "validate" to validate new settings.

@mkanoor with this change, you should be able to revalidate your existing failing ems and have it now update the authentication status.

@Fryguy @blomquisg Please review the commits one by one.  I deleted many redundant tests in the first commit.  (saves 1.5 seconds to run the tests locally :heart_eyes: )

Note, I changed the title and description of this PR to note that I'm only fixing the EMS edit of an existing instance in THIS PR.  I'll look at the EMS new and Host edit/new in a subsequent PR.  EMS validation is most important since it prevents workers from starting if it's invalid.

Please let me know what you think... I limited the scope of this PR to get it right.

Thanks!
cc @brandondunne

jrafanie commented:
cc @himdel @ladas @abonas... please let anyone else know about this change... Many have been bumped by the change #1691 which caused workers to not start the moment an ems no longer has valid authentication.

AparnaKarve commented:
@jrafanie The UI code looks OK to me...however, while testing this noticed that if you click Validate for an existing ems record that failed validation before, in order to fix the validation with the correct credentials, it does not update the ```authentications``` record with a status ```'Ok'```, although the flash message does indicate that ```"Credential validation was successful"```

But if you click on Save to update the ems record, the ```authentications``` record gets correctly updated.

jrafanie commented:
Correct, the validate button only saves an unchanged ems. The use case is an ems that isn't accessible and fails authentication check. You edit the ems and click validate. The validate button with changes to the ems edit form should never save the status since those changes will be checked on save or removed on cancel.

To test locally, you'll need to modify your /etc/hosts file to add an entry for your ems with a bogus ip address. That will enable you to have an unreachable ems that you can then fix in the ui by removing the /etc/hosts entry and edit the ems and click validate.

jrafanie commented:
@AparnaKarve Please ping me if my explanation wasn't clear.  Thanks!

AparnaKarve commented:
@jrafanie I tested this successfully by making the ems unreachable (by tweaking /etc/hosts, per your suggestion), so now on a Validate, the status of the ems *does* change to "Error". 
Likewise, when I make the ems reachable again, Validate changes the status back to "Valid"

Fryguy commented:
@AparnaKarve This is why I suggested to @jrafanie (and I think he left as a TODO comment) to have the UI show "Revalidate" instead of "Validate" for the button until someone makes a change to a different field.  Once they change a field it becomes "Validate".  I think that would be a lot clearer.  Is that possible to do?

AparnaKarve commented:
@Fryguy It should be possible. 
In Angular, it would be a much cleaner implementation since all you need to do is check for "dirty" (changed) fields on the form and reword the button text.

blomquisg commented:
@jrafanie said:
> Did you have a BZ I can link to this issue?

I'm looking for one.  I'm pretty sure there's one regarding Ovirt integration.

jrafanie commented:
:tada: Thanks @blomquisg I'll take it from here.  I'll repush with the right BZ url.  

Was there anything else needed for this PR?

jrafanie commented:
Ok, @blomquisg @AparnaKarve I repushed with the referenced BZ.  Once green, this is ready for final review/merge.

jrafanie commented:
ok, @blomquisg @Fryguy anything else for this PR?  It's all :green_heart:  now

blomquisg commented on vmdb/app/models/authentication.rb:
```diff
@@ -16,6 +16,7 @@ def self.new(*args, &block)
   before_save :set_credentials_changed_on
   after_save :after_authentication_changed
 
+  # TODO: DELETE ME!!!!
```
Yes please!  This belongs in translations

blomquisg commented:
I guess I'm fine with this.  I'm not crazy about returning an array from `authentication_check`.  But, it seems like the only other way to handle this type of situation where you wanna return either `true`, or `false` and an error message is to raise errors during authentication checks.  I'm not super crazy about that notion either.

jrafanie commented:
@blomquisg @Fryguy I agree. How about something like awesome spawn's CommandResult object that could respond to passed?, failed? and message so we can have one place with the meaning of the various results and if you only care about one if these things you can just ask the object. We could then remove the hard coded dictionary UI friendly terms and do a lookup instead.  The downside is authentication check still isn't a predicate method as it used to be.

jrafanie commented:
Plus, the inevitable, is it valid as a string or symbol.

Fryguy commented:
I can see a class or struct defined under Authentication like
Authentication::Status that has these methods.   Should verify credentials
still do what it does now?  It seems the provider specific code would know
how to translate errors better.

I guess my only other thought is that returning a tuple of boolean, reason
is not really uncommon outside of our app,  even in other languages (Elixir
and Go are in my mind right now).

If you did so it, maybe the tuple itself should be more generic.   Just a
Vmdb::Status or something.
On Mar 14, 2015 10:22 AM, "Joe Rafaniello" <notifications@github.com> wrote:

> @blomquisg <https://github.com/blomquisg> @Fryguy
> <https://github.com/Fryguy> I agree. How about something like awesome
> spawn's CommandResult object that could respond to passed?, failed? and
> message so we can have one place with the meaning of the various results
> and if you only care about one if these things you can just ask the object.
> We could then remove the hard coded dictionary UI friendly terms and do a
> lookup instead. The downside is authentication check still isn't a
> predicate method as it used to be.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ManageIQ/manageiq/pull/1990#issuecomment-80506251>.
>

jrafanie commented:
@Fryguy @blomquisg can you review the [last commit](https://github.com/jrafanie/manageiq/commit/3f8c9592da7cc9176281046a30f5bf512491e998), I changed `authentication_check` to return a [boolean result, message]

It changed callers from:
```ruby
    result, details = verify_ems.authentication_check(params[:type], :save => revalidating)
    if result == :valid
      add_flash(_("Credential validation was successful"))
    else
      add_flash(_("Credential validation was not successful: #{details}"), :error)
    end
```

To: (dropped the `== :valid`)

```ruby
    result, details = verify_ems.authentication_check(params[:type], :save => revalidating)
    if result
      add_flash(_("Credential validation was successful"))
    else
      add_flash(_("Credential validation was not successful: #{details}"), :error)
    end
```

I think this simplifies this a great deal.  If this is good, I'll take it for a spin in the UI before I mark this was ready to be merged.

blomquisg commented on vmdb/app/models/ems_kubernetes.rb:
```diff
@@ -40,7 +40,7 @@ def self.event_monitor_class
   def authentication_check
     # TODO: support real authentication using certificates
-    true
+    [true, ""]
```
Is this still the case?  I thought there was a way to validate the Kubernetes "auth".  Maybe not, though...

Fryguy replied:
@jrafanie Can you rebase?  This was changed upstream I thought

Fryguy replied:
OH no wait...gahhh: 
https://github.com/ManageIQ/manageiq/pull/2156

jrafanie replied:
Mwwwhahaha, i'll get my merged first.  Please!!!

blomquisg commented on vmdb/app/models/mixins/authentication_mixin.rb:
```diff
@@ -191,42 +191,44 @@ def authentication_check_types(*args)
     types.to_miq_a.each { |t| self.authentication_check(t, options)}
   end
 
+  # TODO: :valid, :incomplete, and friends shouldn't be littered in here and authentication
```
Can you add a short description here about the return value?  Just want this to be uber clear throughout.  I'm not super crazy about the array return type, but I think it's fine as long as we have a quick blurb here about how it works.

blomquisg commented:
I'm good to go on this after:

1) rename the `authentication_check_no_validation` to something that means that it's checking the authentication but not saving the validation status to the DB

2) a quick comment on around the `authentication_check` method to indicate exactly how the two element array return value is used

jrafanie commented:
> 1) rename the authentication_check_no_validation to something that means that it's checking the authentication but not saving the validation status to the DB

I'll make it private and hide the poorly named method :trophy: 
> 2) a quick comment on around the authentication_check method to indicate exactly how the two element array return value is used

Done.

jrafanie commented:
@blomquisg @Fryguy This is now ready to :ship:, any thing else?  Hopefully tests still pass ;-)

jrafanie commented:
:bow:  :tada: :heart: Thank you!  That was a team effort.
